_posts/2018-02-20--inception-pipelines-pt1.markdown

```diff
@@ -0,0 +1,99 @@
+
+1. An AWS Account to create the pipeline in.
+1. The AWS CLI installed and configured with [access credentials](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html).
+1. CodeCommit [git credentials configured](https://docs.aws.amazon.com/codecommit/latest/userguide/setting-up-https-unixes.html#setting-up-https-unixes-credential-helper).
+
+While not strictly required, a passing familarity of Bash Shell scripts, Git, JSON & YAML, and CloudFormation templates will make understanding everything easier.
+
+## How it all works
+
+At a high-level, the Inception Pipeline works by executing a CloudFormation template which then creates a CodeCommit repository, a CodePipeline pipeline and a few other supporting resources. The first non-source action in the pipeline is a [CloudFormation Deployment Action](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/continuous-delivery-codepipeline.html). This blog post won't dive deeply into the CloudFormation template (I'll leave that as an exercise for you dear reader). Instead I'll just discuss the really juicy bits.
+
+The secret-sauce to the Inception Pipeline is the using the same CloudFormation stackname in the CLI/console invocation and the CodePipeline Action (CloudFormation parameter ```StageAdministerPipelineStackName```). Once the initial CloudFormation stack has been created, the CodePipeline action below maintains the pipeline.
+
+```yaml
+- Name: 'AdministerPipeline'
+    Actions:
+    - Name: 'AdministerPipeline'
+        ActionTypeId:
+            Category: 'Deploy'
+            Owner: 'AWS'
+            Provider: 'CloudFormation'
+            Version: '1'
+        Configuration:
+            ActionMode: 'REPLACE_ON_FAILURE' 
+            Capabilities: 'CAPABILITY_NAMED_IAM'
+            RoleArn: !GetAtt [CloudFormationDeployActionRole, Arn]
+            StackName: !Ref StageAdministerPipelineStackName
+            TemplateConfiguration: !Join ['', [!Ref RepositoryName, 'Source', '::aws_seed.json']]
+            TemplatePath: !Join ['', [!Ref RepositoryName, 'Source', '::aws_seed.yml']]
+        InputArtifacts:
+            - Name: !Join ['', [!Ref RepositoryName, 'Source']]
+        RunOrder: '1'
+```
+
+The CodePipeline is also configured to restart execution if it updates itself (```RestartExecutionOnUpdate: true```). This means when you add new stages, actions, etc, the pipeline will run the above action, realise its changed and then restart itself.
+
+While not strictly necessary, I would recommend that:
+
+1. The ```AdministerPipeline``` stage is only used for the ```AdministerPipeline``` action, i.e. don't add any other actions in.
+2. The ```AdministerPipeline``` stage is the first stage after the ```Source``` stage. This allows the pipeline to update itself before anything else runs.
+
+## Where do I get the seed files
+
+Go grab a copy from the [GitHub repository](https://github.com/MechanicalRock/InceptionPipeline/part1). I'll wait while you do.
```

Comment:
> I'll wait while you do.
New paragraph - stands out better and gets me ROFLing :)

Reply:
Done